vpr/src/route/route_tree_type.h

```diff
@@ -1,60 +1,210 @@
- *                 so the net pin index guarantees an unique identification  *
- *                 for each SINK rt_node. For non-SINK nodes and for SINK    *
- *                 nodes with no associated net pin index, (i.e. special     *
- *                 SINKs like the source of a clock tree which do not        *
- *                 correspond to an actual netlist connection), the value    *
- *                 for this member should be set to OPEN (-1).               *
- * C_downstream:  Total downstream capacitance from this rt_node.  That is,  *
- *                the total C of the subtree rooted at the current node,     *
- *                including the C of the current node.                       *
- * R_upstream:  Total upstream resistance from this rt_node to the net       *
- *              source, including any device_ctx.rr_nodes[].R of this node.  *
- * Tdel:  Time delay for the signal to get from the net source to this node. *
- *        Includes the time to go through this node.                         */
-struct t_rt_node {
-    union {
-        t_linked_rt_edge* child_list;
-        t_rt_node* next;
-    } u;
-    t_rt_node* parent_node;
-    short parent_switch;
+    RouteTreeNode(RRNodeId inode, RRSwitchId parent_switch, vtr::optional<RouteTreeNode&> parent, vtr::optional<RouteTree&> root);
+
+    RRNodeId inode;
+    RRSwitchId parent_switch;
+    vtr::optional<RouteTreeNode&> parent;
+    vtr::optional<RouteTree&> root;
     bool re_expand;
-    int inode;
     int net_pin_index;
     float C_downstream;
     float R_upstream;
     float Tdel;
+    std::list<RouteTreeNode> _child_nodes;
+
+    /* Copy child to the back of _child_nodes.
+     * Returns a reference to the added node. */
+    RouteTreeNode& add_child(const RouteTreeNode&);
+
+    /* Copy child to the front of _child_nodes.
+     * Returns a reference to the added node. */
+    RouteTreeNode& add_child_front(const RouteTreeNode&);
+
+    /* Emplace child to the back of _child_nodes.
+     * For best performance, call with constructor args
+     * (will construct the node in the parent's list directly and save a copy) */
+    template<class... Args>
+    RouteTreeNode& emplace_child(Args&&... args);
+
+    /* Emplace child to the front of _child_nodes.
+     * For best performance, call with constructor args
+     * (will construct the node in the parent's list directly and save a copy) */
+    template<class... Args>
+    RouteTreeNode& emplace_child_front(Args&&... args);
+
+    /* Remove child node by value. O(N) operation. */
+    void remove_child(const RouteTreeNode&);
+
+    /* Remove child node by list iterator state. O(1) operation. */
+    void remove_child(std::list<RouteTreeNode>::iterator&);
+
+    /* Iterate through child nodes and remove if p returns true.
+     * Best way to use is through a lambda which takes "child_node" as an argument. */
+    void remove_child_if(const std::function<bool(RouteTreeNode&)>&);
```

Comment:
Deleting things while iterating can become confusing quickly and requires one to use the old iterator syntax rather than `for(x : xs) ...` Here you can see how smooth `remove_if` is as an interface.

https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/1ea38b7f57394bead0ac4208855f062252a79724/vpr/src/route/route_timing.cpp#L1569-L1581